DataStructures/HashTable/HashTable.cs

```diff
@@ -32,62 +38,81 @@ public bool Contains(object key)
         public void Add(object key, object value)
         {
             int index = Hash(key);
-            if (index <= _array.Length - 1 && _array[index] != null)
+            if (Contains(key))
             {
                 throw new DuplicateNameException();
             }
 
-            if (index > _array.Length - 1)
+            if (CheckForFull(index))
             {
-                Copy(_array, index);
+                _array = Resize(index + 1);
             }
 
-            _array[index] = value;
+            _array[index] = new ArrayValue();
+            _array[index].Value = value;
         }
 
-        public void Copy(object[] array, int index)
+        private bool CheckForFull(int index)
         {
-            object[] temp = new object[index + 1];
+            if (index > _array.Length)
+                return true;
+            return false;
+
+        }
+
+        private ArrayValue[] Resize(int size)
+        {
+            ArrayValue[] temp = new ArrayValue[size];
             _array.CopyTo(temp, 0);
-            _array = temp;
+            return temp;
         }
 
         public object this[object key]
         {
-            get => _array[Hash(key)];
-
+            get => CheckForFull(Hash(key)) ? null : _array[Hash(key)].Value;
             set
             {
-                int index = Hash(key);
-                if (index > _array.Length - 1)
+                if (value == null)
+                {
+                    Remove(key);
+                }
+
+                else if (Contains(key))
                 {
-                    if (value == null)
-                    {
-                        return;
-                    }
+                    _array[Hash(key)].Value = value;
+                }
 
-                    Copy(_array, index);
+                else
+                {
+                    Add(key, value);
                 }
+            }
+        }
 
-                _array[index] = value;
+        public void Remove(object key)
+        {
+            if (Contains(key))
+            {
+                _array[Hash(key)] = null;
             }
+
         }
 
         public bool TryGet(object key, out object value)
         {
             int index = Hash(key);
-            if (index > _array.Length - 1)
+            if (CheckForFull(index))
             {
                 value = null;
                 return false;
             }
 
-            if (_array[index] == null)
+            if (!Contains(key))
             {
                 throw new NullReferenceException();
```

Comment:
I know this is done according to the requirements and it's done right. But it's confusing we have two behaviors when the hashtable doesn't contain an item. Isn't it? Let's not throw but make it like above (return false).

Reply:
Okay